File: src/services/audit/detectors/ad/advanced/other/ntlm-relay-opportunity.ts
/**
 * NTLM Relay Opportunity Detector
 * Check for NTLM relay opportunity
 */

import { ADDomain } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';

export function detectNtlmRelayOpportunity(domain: ADDomain | null, includeDetails: boolean): Finding {
  if (!domain) {
    return {
      type: 'NTLM_RELAY_OPPORTUNITY',
      severity: 'medium',
      category: 'advanced',
      title: 'NTLM Relay Opportunity',
      description: 'Unable to check LDAP signing configuration.',
      count: 0,
    };
  }

  const ldapSigningRequired = (domain as any).ldapSigningRequired;
  const channelBindingRequired = (domain as any).channelBindingRequired;

  const isVulnerable = !ldapSigningRequired || !channelBindingRequired;

  return {
    type: 'NTLM_RELAY_OPPORTUNITY',
    severity: 'medium',
    category: 'advanced',
    title: 'NTLM Relay Opportunity',
    description: 'LDAP signing or channel binding not enforced. Enables NTLM relay attacks.',
    count: isVulnerable ? 1 : 0,
    affectedEntities: includeDetails && isVulnerable ? [domain.dn] : undefined,
  };
}

File: internal/audit/detectors/ad/advanced/other/ntlm-relay-opportunity.go
package other

import (
	"context"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// NtlmRelayOpportunityDetector detects NTLM relay opportunities
type NtlmRelayOpportunityDetector struct {
	audit.BaseDetector
}

// NewNtlmRelayOpportunityDetector creates a new detector
func NewNtlmRelayOpportunityDetector() *NtlmRelayOpportunityDetector {
	return &NtlmRelayOpportunityDetector{
		BaseDetector: audit.NewBaseDetector("NTLM_RELAY_OPPORTUNITY", audit.CategoryAdvanced),
	}
}

// Detect executes the detection
func (d *NtlmRelayOpportunityDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	if data.DomainInfo == nil {
		return []types.Finding{{
			Type:        d.ID(),
			Severity:    types.SeverityMedium,
			Category:    string(d.Category()),
			Title:       "NTLM Relay Opportunity",
			Description: "Unable to check LDAP signing configuration.",
			Count:       0,
		}}
	}

	ldapSigningRequired := data.DomainInfo.LDAPSigningRequired
	channelBindingRequired := data.DomainInfo.ChannelBindingRequired

	isVulnerable := !ldapSigningRequired || !channelBindingRequired

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityMedium,
		Category:    string(d.Category()),
		Title:       "NTLM Relay Opportunity",
		Description: "LDAP signing or channel binding not enforced. Enables NTLM relay attacks.",
		Count:       0,
	}

	if isVulnerable {
		finding.Count = 1
		if data.IncludeDetails {
			finding.AffectedEntities = []types.AffectedEntity{{
				Type:           "domain",
				SAMAccountName: data.DomainInfo.DN,
			}}
		}
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewNtlmRelayOpportunityDetector())
}
